sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -174,8 +106,110 @@ public void executeDdl(String sqlStatement) throws ParseException {
+      for (Map.Entry<String, List<Method>> entry : methods.entrySet()) {
+        for (Method method : entry.getValue()) {
+          jdbcConnection.getCurrentSchemaPlus().add(entry.getKey(), UdfImpl.create(method));
+        }
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder addSchema(String name, TableProvider tableProvider) {
+      jdbcConnection.setSchema(name, tableProvider);
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setCurrentSchema(String name) {
+      try {
+        jdbcConnection.setSchema(name);
+      } catch (SQLException e) {
+        throw new RuntimeException(e);
+      }
+
+      return this;
+    }
+
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<?> clazz, String method) {
+      jdbcConnection.getCurrentSchemaPlus().add(functionName, UdfImpl.create(clazz, method));
+
+      return this;
+    }
+
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
+      return registerUdf(functionName, clazz, BeamSqlUdf.UDF_METHOD);
+    }
+
+    public BeamSqlEnvBuilder registerUdf(String functionName, SerializableFunction sfn) {
+      return registerUdf(functionName, sfn.getClass(), "apply");
+    }
+
+    /**
+     * Register a UDAF function which can be used in GROUP-BY expression. See {@link
+     * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
+     */
+    public BeamSqlEnvBuilder registerUdaf(String functionName, Combine.CombineFn combineFn) {
+      jdbcConnection.getCurrentSchemaPlus().add(functionName, new UdafImpl(combineFn));
+
+      return this;
+    }
+
+    /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
+    public BeamSqlEnvBuilder loadUdfUdafFromProvider() {
+      ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
+          .forEach(
+              ins -> {
+                ins.getBeamSqlUdfs().forEach((udfName, udfClass) -> registerUdf(udfName, udfClass));
+                ins.getSerializableFunctionUdfs()
+                    .forEach((udfName, udfFn) -> registerUdf(udfName, udfFn));
+                ins.getUdafs().forEach((udafName, udafFn) -> registerUdaf(udafName, udafFn));
+              });
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder loadBeamBuiltinFunctions() {
+      for (BeamBuiltinFunctionProvider provider :
+          ServiceLoader.load(BeamBuiltinFunctionProvider.class)) {
+        registerBuiltinUdf(provider.getBuiltinMethods());
+      }
+
+      return this;
+    }
+
+    public BeamSqlEnvBuilder setPlannerName(String name) {
+      plannerName = name;
+      return this;
+    }
+
+    public BeamSqlEnv build() {
+      if (plannerName.equals("CalcitePlanner")) {
+        return new BeamSqlEnv(jdbcConnection, new CalciteQueryPlanner(jdbcConnection));
```

Comment:
nit:I think the pattern is usually to pass the builder into the constructor and then extract builder fields inside the constructor, this way when changing the builder interface you need to change fewer places